docs/microbenchmark-design-guidelines.md

```diff
@@ -0,0 +1,557 @@
+    }
+}
+```
+
+In this case, there should be one benchmark for compression and another for the decompression.
+
+### No Side-Effects
+
+Let's consider following benchmark:
+
+```cs
+List<int> _numbers = new List<int>();
+
+[Benchmark]
+public void Add() => _numbers.Add(12345);
+```
+
+In this particular benchmark, the list is growing with every benchmark invocation. `List<T>` is internally using an `Array` to store all the elements. When the array is not big enough to store one more element, a two times bigger array is allocated and all elements are copied from the old array to the new one. It means that every next `Add` operation takes more time. We might also get `OutOfMemoryException` at some point in time.
+
+**Benchmarks should not have any side effects**. 
+
+### Dead Code Elimination
+
+To prevent from [dead code elimination](https://en.wikipedia.org/wiki/Dead_code_elimination) BenchmarkDotNet consumes the result returned from a benchmark and writes it to a `volatile` field ([code](https://github.com/dotnet/BenchmarkDotNet/blob/94863ab4d024eca04d061423e5aad498feff386b/src/BenchmarkDotNet/Engines/Consumer.cs)).
+
+The only thing that you need to remember is to **return the result from the benchmark**. Even if you think that as of today the JIT is not going to eliminate the code.
+
+### Loops
+
+BenchmarkDotNet does not require the user to provide the number of **invocations per iteration**. This value is determined by BenchmarkDotNet during the Pilot Experiment Stage, based on the `IterationTime` setting.
+
+It means, that you don't have to write a loop inside the benchmark and repeat the operation many times. The harness is going to do this for you.
+
+```cs
+[Benchmark]
+[Arguments("12345")]
+public int ParseOk(string value) => int.Parse(value);
+
+[Benchmark]
+[Arguments("12345")]
+public int ParseNotOk(string value)
+{
+    int result = 0;
+    for (int i = 0; i < 1000; i++)
+    {
+        result += int.Parse(value);
+    }
+    return result;
+}
+```
+
+By relying on the BDN mechanism you are going to avoid loop alignment issues. BenchmarkDotNet handles that on its own side when your benchmark does not have a loop.
+
+### Method Inlining
+
+BenchmarkDotNet prevents from inlining the benchmarked method by wrapping it into a delegate (delegates can not be inlined as of today). The cost of delegate invocation is excluded by running separate run for Overhead calculation.
+
+The benchmark methods don't need to have `[MethodImpl(MethodImplOptions.NoInlining)]` attribute applied.
+
+## Summary
+
+* Choose a good name because you won't be able to change it after the first results are uploaded to the Reporting System.
+* Move initialization logic to a Setup method. Don't forget to clean up the resources in a corresponding Cleanup method.
+* If your benchmark needs input data, it should be always exactly the same data. Use `ValuesGenerator` to generate the data.
+* You should have one test case per one code path. Not more.
```

Comment:
this feels a little ambiguous between "benchmarks should do one thing" and "one benchmark for each code path I am trying to test". You are saying the second here - maybe just be clearer?